### PR TITLE
Implement stream states for OpenSL ES streams

### PR DESCRIPTION
**Pull request by dturner**
Adding tests and states for OpenSL ES input and output streams.

### CODE REVIEW

File: src/opensles/AudioOutputStreamOpenSLES.cpp [hunk 1]
```diff
@@ -187,98 +189,83 @@ Result AudioOutputStreamOpenSLES::close() {
...
-            processBufferCallback(mSimpleBufferQueueInterface);
-            setState(StreamState::Starting);
-        }
-        return result;
+    setState(StreamState::Starting);
+    Result result = setPlayState(SL_PLAYSTATE_PLAYING);
+    if (result == Result::OK) {
+        setState(StreamState::Started);
+        processBufferCallback(mSimpleBufferQueueInterface);
     }
```
Comment (philburk) [inline src/opensles/AudioOutputStreamOpenSLES.cpp]: restore original state on error

File: src/opensles/AudioStreamOpenSLES.cpp [hunk 2]
```diff
@@ -248,3 +250,29 @@ int64_t AudioStreamOpenSLES::getFramesProcessedByServer() const {
...
+        timeoutNanoseconds -= durationNanos;
+
+        state = getState();
+    }
+    if (nextState != nullptr) {
+        *nextState = state;
+    }
+
+    return (state == currentState) ? Result::ErrorTimeout : Result::OK;
+}
```
Comment (philburk) [inline src/opensles/AudioStreamOpenSLES.cpp]: add newline

File: src/opensles/AudioStreamOpenSLES.cpp [hunk 3]
```diff
@@ -248,3 +250,29 @@ int64_t AudioStreamOpenSLES::getFramesProcessedByServer() const {
     int64_t framesProcessed = millis64 * getSampleRate() / kMillisPerSecond;
     return framesProcessed;
 }
+
+Result AudioStreamOpenSLES::waitForStateChange(StreamState currentState,
+                                                     StreamState *nextState,
+                                                     int64_t timeoutNanoseconds) {
+    LOGD("AudioStreamOpenSLES::waitForStateChange()");
+
+    if (getState() == StreamState::Closed) return Result::ErrorClosed;
```
Comment (philburk) [inline src/opensles/AudioStreamOpenSLES.cpp]: Maybe does not need to be an error. Just let it run.

File: tests/testStreamStates.cpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+    AudioStreamBuilder mBuilder;
+    AudioStream *mStream = nullptr;
+    static constexpr int kTimeoutInNanos = 100 * kNanosPerMillisecond;
+
+};
+
+TEST_F(StreamStates, OutputStreamStateIsOpenAfterOpening){
+    openStream();
+    StreamState next;
```
Comment (philburk) [inline tests/testStreamStates.cpp]: Initialize next to something bogus so we know it changes.

File: src/opensles/AudioInputStreamOpenSLES.cpp [hunk 5]
```diff
@@ -194,97 +195,75 @@ Result AudioInputStreamOpenSLES::close() {
...
-    } else {
-        Result result = setRecordState(SL_RECORDSTATE_RECORDING);
-        if (result == Result::OK) {
-            // Enqueue the first buffer so that we have data ready in the callback.
-            enqueueCallbackBuffer(mSimpleBufferQueueInterface);
-            setState(StreamState::Starting);
-        }
-        return result;
+    LOGD("AudioInputStreamOpenSLES::requestStart()");
+    if (mState == StreamState::Closed)return Result::ErrorClosed;
```
[the remark this answers is not in the document]
Comment (dturner) [reply src/opensles/AudioInputStreamOpenSLES.cpp]: done

File: src/opensles/AudioInputStreamOpenSLES.cpp [hunk 6]
```diff
@@ -194,97 +195,75 @@ Result AudioInputStreamOpenSLES::close() {
...
+
+    setState(StreamState::Starting);
+    Result result = setRecordState(SL_RECORDSTATE_RECORDING);
+    if (result == Result::OK) {
+        // Enqueue the first buffer so that we have data ready in the callback.
+        setState(StreamState::Started);
+        enqueueCallbackBuffer(mSimpleBufferQueueInterface);
     }
```
[the remark this answers is not in the document]
Comment (dturner) [reply src/opensles/AudioInputStreamOpenSLES.cpp]: done

File: src/opensles/AudioInputStreamOpenSLES.cpp [hunk 7]
```diff
@@ -194,97 +195,75 @@ Result AudioInputStreamOpenSLES::close() {
...
-        return result;
+    LOGD("AudioInputStreamOpenSLES::requestPause()");
+    if (mState == StreamState::Closed) return Result::ErrorClosed;
+
+    setState(StreamState::Pausing);
+    Result result = setRecordState(SL_RECORDSTATE_PAUSED);
+    if (result == Result::OK) {
+        mPositionMillis.reset32(); // OpenSL ES resets its millisecond position when paused.
+        setState(StreamState::Paused);
     }
```
[the remark this answers is not in the document]
Comment (dturner) [reply src/opensles/AudioInputStreamOpenSLES.cpp]: done

File: src/opensles/AudioInputStreamOpenSLES.cpp [hunk 8]
```diff
@@ -194,97 +195,75 @@ Result AudioInputStreamOpenSLES::close() {
...
-        }
-        return result;
+    setState(StreamState::Stopping);
+
+    Result result = setRecordState(SL_RECORDSTATE_STOPPED);
+    if (result == Result::OK) {
+        mPositionMillis.reset32(); // OpenSL ES resets its millisecond position when stopped.
+        setState(StreamState::Stopped);
     }
+    return result;
```
[the remark this answers is not in the document]
Comment (dturner) [reply src/opensles/AudioInputStreamOpenSLES.cpp]: done

File: src/opensles/AudioOutputStreamOpenSLES.cpp (hunk 1 above)
Comment (dturner) [reply src/opensles/AudioOutputStreamOpenSLES.cpp]: done

File: src/opensles/AudioStreamOpenSLES.cpp (hunk 2 above)
Comment (dturner) [reply src/opensles/AudioStreamOpenSLES.cpp]: done

File: src/opensles/AudioStreamOpenSLES.cpp (hunk 3 above)
Comment (dturner) [reply src/opensles/AudioStreamOpenSLES.cpp]: done

File: tests/testStreamStates.cpp (hunk 4 above)
Comment (dturner) [reply tests/testStreamStates.cpp]: done

File: src/opensles/AudioStreamOpenSLES.cpp (hunk 3 above)
Comment (dturner) [reply src/opensles/AudioStreamOpenSLES.cpp]: and test updated to test for `Result::OK` when stream is closed and `waitForStateChange` is called.

**Pull request merged**